### PR TITLE
[Spec] Added more specs to the `coin` and `account` modules

### PR DESCRIPTION
**Pull request by junkil-park**
### Description
- this PR completes many TODOs in the specs of coin and account modules.
- added more specs and helper spec functions to `ed25519` and `multi_ed25519` modules
- enhanced the spec of `type_info` module
  - TODO: the spec of `chain_id()` is blocked by the bit operation support.

### Test Plan
`move prove`

### CODE REVIEW

File: aptos-move/framework/aptos-framework/sources/account.spec.move [hunk 1]
```diff
@@ -81,8 +81,24 @@ spec aptos_framework::account {
     }
 
     spec assert_valid_signature_and_get_auth_key(scheme: u8, public_key_bytes: vector<u8>, signature: vector<u8>, challenge: &RotationProofChallenge): vector<u8> {
-        // TODO: complex aborts conditions.
-        pragma aborts_if_is_partial;
+        aborts_if scheme == ED25519_SCHEME && len(public_key_bytes) != ed25519::PUBLIC_KEY_NUM_BYTES;
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/account.spec.move:84]: cc @alinush for verifying these conditions

File: aptos-move/framework/move-stdlib/sources/option.move [hunk 2]
```diff
@@ -189,6 +189,7 @@ module std::option {
     }
     spec swap_or_fill {
         pragma opaque;
+        aborts_if false;
```
Comment (movekevin) [inline aptos-move/framework/move-stdlib/sources/option.move:192]: Why add this?

**Review by movekevin — APPROVED** [review]

File: aptos-move/framework/move-stdlib/sources/option.move (hunk 2 above)
Comment (junkil-park) [reply aptos-move/framework/move-stdlib/sources/option.move:192]: This is not turning off the abort check. This asserts that this function should not abort.

File: aptos-move/framework/aptos-framework/sources/coin.spec.move [hunk 3]
```diff
@@ -214,14 +214,13 @@ spec aptos_framework::coin {
...
     spec register<CoinType>(account: &signer) {
-        // TODO: Add the abort condition about `type_info::type_of`
-        pragma aborts_if_is_partial;
         let account_addr = signer::address_of(account);
         let acc = global<account::Account>(account_addr);
         aborts_if acc.guid_creation_num + 2 > MAX_U64;
         aborts_if exists<CoinStore<CoinType>>(account_addr);
         aborts_if !exists<account::Account>(account_addr);
         ensures exists<CoinStore<CoinType>>(account_addr);
+        aborts_if !type_info::spec_is_struct<CoinType>();
```
Comment (meng-xu-cs) [inline aptos-move/framework/aptos-framework/sources/coin.spec.move:223]: Move this line up before `ensures`

**Review by meng-xu-cs — COMMENTED** [review]
Leaving the `ed25519` spec to crypto experts, the rest looks good to me.

File: aptos-move/framework/aptos-framework/sources/coin.spec.move (hunk 3 above)
Comment (junkil-park) [reply aptos-move/framework/aptos-framework/sources/coin.spec.move:222]: Done!

File: aptos-move/framework/aptos-framework/sources/account.spec.move [hunk 4]
```diff
@@ -81,8 +81,24 @@ spec aptos_framework::account {
...
+        aborts_if scheme == ED25519_SCHEME && len(signature) != ed25519::SIGNATURE_NUM_BYTES;
+        aborts_if scheme == ED25519_SCHEME && !ed25519::spec_signature_verify_strict_t(
+            ed25519::Signature { bytes: signature },
+            ed25519::UnvalidatedPublicKey { bytes: public_key_bytes },
+            challenge
+        );
+        aborts_if scheme == MULTI_ED25519_SCHEME && (
```
Comment (alinush) [inline aptos-move/framework/aptos-framework/sources/account.spec.move:91]: The fact that the spec of `account.move` has to be aware of all of these digital-signature related details (e.g., PK size), seems to break any reasonable notion of modularity.

Can we avoid this?

After all, these conditions were already encoded in the specs of `ed25519.move` and `multi_ed25519.move`

File: aptos-move/framework/aptos-framework/sources/account.spec.move [hunk 5]
```diff
@@ -116,11 +132,41 @@ spec aptos_framework::account {
         account_public_key_bytes: vector<u8>,
         recipient_address: address
     ) {
-        // TODO: complex aborts conditions.
-        pragma aborts_if_is_partial;
         let source_address = signer::address_of(account);
+        let account_resource = global<Account>(source_address);
```
Comment (alinush) [inline aptos-move/framework/aptos-framework/sources/account.spec.move:136]: This spec also looks correct, but I still have the same modularity concerns [from above](https://github.com/aptos-labs/aptos-core/pull/5832/files#r1046629266)

**Review by alinush — DISMISSED** [review]
Can we make the specs more modular? (see comments)

File: aptos-move/framework/aptos-framework/sources/account.spec.move (hunk 4 above)
Comment (junkil-park) [reply aptos-move/framework/aptos-framework/sources/account.spec.move]: @alinush , I made it modular. Please let me know how that looks to you. Thank you!

File: aptos-move/framework/aptos-framework/sources/account.spec.move (hunk 5 above)
Comment (junkil-park) [reply aptos-move/framework/aptos-framework/sources/account.spec.move:133]: @alinush , I made this modular too. Please let me know what you think.

**Review by rahxephon89 — APPROVED** [review]

**Pull request merged**